### ISSUE TITLE
Cool project! I've been trying to work on an AIL/32 HDA driver.

### ISSUE THREAD

**Issue by volkertb** [PluMGMK/hmidrv_hdaudio#1]
Hi @PluMGMK,

Through some (old) discussion threads in the freedos-dev mailing list, I stumbled upon this project.

Really cool how you got this to work, at least with Rayman.

I've been trying to implement the same idea, but as an AIL/32 driver (the protected mode version of the Miles AIL 2.x drivers).

John Miles open-sourced those drivers (along with the real mode .adv counterparts) years ago, and GitHub user @Wohlstand [had already ported the AIL/32 driver sources to a modern open-source toolchain](https://github.com/Wohlstand/ail32-sandbox).

I haven't gotten very far yet. But with some help from people in the [VCFED forums](https://forum.vcfed.org/), I managed to at least figure out how to link C code into an AIL/32 driver. I had to keep a basic shim in assembly and link that with C sources. For some reason, I've been unable to build an AIL/32 driver DLL in pure C. Also, I don't have access to the standard C library in this way. Probably because AIL/32 basically uses the OS/2 DLL format, since the Watcom toolchain already supported that compilation target. But when compiling to the OS/2 DLL target, the standard C runtime is not available, which I believe is part of the reason why I don't have access to the standard C library inside the driver sources.

I was thinking about maybe using [picolibc](https://github.com/picolibc/picolibc) to have access to a lightweight C library in the driver source, in case I would have to rely on a lot of high-level C stuff. Likely I will, since it's my intention to use the Linux HDA sound driver sources in this project.

Sorry, I'm rambling on about my AIL/32 HDA project. :sweat_smile:

Long story short, I would love to learn from you. Perhaps you could give me some pointers in getting this to work. I noticed you already had the Miles drivers on the TODO list of the README of your hmidrv_hdaudio project.

Anyway, I'm curious to read what you think.

By the way, great work on that high-color Windows 3.x driver as well! Nice to see such modes finally being unlocked again on newer hardware.

Also perhaps useful to share: I'm involved with @crazii's [SBEMU](https://github.com/crazii/SBEMU) project. A DOS Sound Blaster emulator for newer PCs. Should you be interested, that project could always use more enthusiasts willing to help improve it.

Regardless, have a Merry Christmas and a Happy and Healthy New Year!

**Comment by PluMGMK** [comment]
> Hi @PluMGMK,

Hello, Happy Christmas! :smiley: 

> Through some (old) discussion threads in the freedos-dev mailing list, I stumbled upon this project.
> 
> Really cool how you got this to work, at least with Rayman.

Thank you! :smiley: 

> I've been trying to implement the same idea, but as an AIL/32 driver (the protected mode version of the Miles AIL 2.x drivers).
> 
> John Miles open-sourced those drivers (along with the real mode .adv counterparts) years ago, and GitHub user @Wohlstand [had already ported the AIL/32 driver sources to a modern open-source toolchain](https://github.com/Wohlstand/ail32-sandbox).

Ah yes, I think I had noticed that, which is why I put the AIL version on my TODO list. I guess I never worked on it because I don't really have a game to test it with, at least none that I know as well as Rayman!

> I haven't gotten very far yet. But with some help from people in the [VCFED forums](https://forum.vcfed.org/), I managed to at least figure out how to link C code into an AIL/32 driver. I had to keep a basic shim in assembly and link that with C sources. For some reason, I've been unable to build an AIL/32 driver DLL in pure C. Also, I don't have access to the standard C library in this way. Probably because AIL/32 basically uses the OS/2 DLL format, since the Watcom toolchain already supported that compilation target. But when compiling to the OS/2 DLL target, the standard C runtime is not available, which I believe is part of the reason why I don't have access to the standard C library inside the driver sources.

Interesting… I suppose since it's a small loadable module they don't want to load it up with C runtime stuff that's likely to be already present in the program that's loading it.

> I was thinking about maybe using [picolibc](https://github.com/picolibc/picolibc) to have access to a lightweight C library in the driver source, in case I would have to rely on a lot of high-level C stuff. Likely I will, since it's my intention to use the Linux HDA sound driver sources in this project.

Hmm… If you're planning to adapt source from the Linux kernel, then as far as I know that doesn't rely on a normal C runtime, but on a bunch of C frameworks defined in the kernel itself. It can probably be adapted to use a normal C runtime like picolibc though.

Where you might run into trouble is actually finding and communicating with the PCI device. IIRC you'll need to do the following, probably with a bit of assembly involved:
* Search the PCI bus for a HD Audio device, either using the PCI BIOS through `int 1Ah`, or directly accessing PCI Configuration Space using Port I/O.
* Read the device's Base Address Register in PCI Configuration Space, again either using port I/O or the PCI BIOS.
* Map it into your address space using DPMI, with [function `int 31h AX=0800h`](https://fd.lod.bz/rbil/interrup/dos_extenders/310800.html).

From there you'll get a pointer to the device's configuration registers, which you can probably use with C code adapted from Linux. But you'll also need to allocate some physical memory for the CORB and RIRB, and "somehow" map it into your virtual address space while also knowing its physical address so you can set it correctly in the configuration registers.
* IIRC in the driver I did this rather clunkily, by making a Real-Mode call to the XMS driver to allocate and lock some memory, and then using `int 31h AX=0800h` to map it into my address space.
* If I were doing it over again, I would probably [use `int 4Bh`](https://fd.lod.bz/rbil/interrup/io_disk/4b8103.html#table-03221) to allocate the physical memory instead.
* If `int 4Bh` is not present, then as a fallback you can just allocate some memory as normal (using `malloc` or `int 31h AX=0501h`) and assume that physical and virtual addresses are identical.

Honestly, I don't have much experience writing C drivers, or studying Linux kernel sources / drivers. I guess I'm still working my way up from 80s-/early-90s-level driver development! :smile: 

You've probably already looked at the [relevant OSdev wiki page](https://wiki.osdev.org/HD_Audio) but I'll leave it here just in case. I frequently consulted it while working on this!

> Sorry, I'm rambling on about my AIL/32 HDA project. 😅
> 
> Long story short, I would love to learn from you. Perhaps you could give me some pointers in getting this to work. I noticed you already had the Miles drivers on the TODO list of the README of your hmidrv_hdaudio project.
> 
> Anyway, I'm curious to read what you think.
> 
> By the way, great work on that high-color Windows 3.x driver as well! Nice to see such modes finally being unlocked again on newer hardware.

Thanks! I was working on that thing on-and-off for over a year, but I decided it was finally "good enough" to reveal it. For a long time I didn't want to put it out there in case I might lose interest and leave it unfinished. :sweat_smile: 

> Also perhaps useful to share: I'm involved with @crazii's [SBEMU](https://github.com/crazii/SBEMU) project. A DOS Sound Blaster emulator for newer PCs. Should you be interested, that project could always use more enthusiasts willing to help improve it.

Ah yes, I think I've seen that before. I would love to adapt it into a VxD so that it could be used with Windows 3.1, but I guess since it's GPL I can't really combine it with the Win16 DDK (which is definitely not GPL!) :sweat_smile: 

> Regardless, have a Merry Christmas and a Happy and Healthy New Year!

Many happy returns! :christmas_tree:
